### PR TITLE
prov/verbs: Use RDMA CM SIDR message exchanges for shared XRC QP connections

### PR DESCRIPTION
**Pull request by swelch**
This is a series of patches that implement using RDMA CM SIDR message exchanges to setup virtual/shared XRC QP connections. This avoids the necessity of creating temporary hardware resources when setting up the shared connections. With these changes XRC connection setup becomes faster than RC when the PPN count is greater than 4 PPN, and has up to a 70% improvement in connection setup time over the existing XRC connection setup implementation.

### CODE REVIEW

**Comment by swelch** [comment]
@shefty -  I've done some extensive testing with this PR at scale, including lost SIDR response message recovery.

File: prov/verbs/src/fi_verbs.h [hunk 1]
```diff
@@ -289,6 +289,14 @@ struct fi_ibv_eq {
 		 * consider using an internal PEP listener for handling the
 		 * internally processed reciprocal connections. */
 		uint16_t		pep_port;
+
+		/* SIDR request/responses are a two-way handshake; therefore,
+		 * we maintain an RB tree of SIDR accept responses, so that if
+		 * a response is lost, the subsequent retried request can be
+		 * detected and the original accept response resent. Note, that
+		 * reject responses can be passed to RXM and will be rejected
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.h]: nit: I think this should be 'rejected requests' instead of responses.

**Comment by swelch** [comment]
> A few minor comments. Since SIDR uses a 2-way handshake, how does the responder know when it is okay to send data to the peer? This is the purpose behind the RTU message. Does this problem not apply here, since you're using 2-way handshakes from both sides?

Right, it does not apply. For XRC the unidirectional nature of the connections requires us to create connections in each direction. This allows the accept response of the reverse connection which is sent by the original active side to the original passive side to serve the purpose of the composite connection RTU.

File: prov/verbs/src/verbs_domain_xrc.c [hunk 2]
```diff
@@ -291,6 +301,23 @@ void fi_ibv_sched_ini_conn(struct fi_ibv_ini_shared_conn *ini_conn)
...
+		ret = rdma_migrate_id(ep->base_ep.id,
+				      ep->base_ep.eq->channel);
+		if (ret) {
+			VERBS_WARN(FI_LOG_EP_CTRL,
+				   "Failed to migreate active CM ID %d\n", ret);
+			goto err;
+		}
+
+		peer_addr = rdma_get_local_addr(ep->base_ep.id);
+		if (peer_addr)
```
[the remark this answers is not in the document]
Comment (swelch) [reply prov/verbs/src/verbs_domain_xrc.c]: Thanks, I will update the associated commit.

File: prov/verbs/src/verbs_ep.c [hunk 3]
```diff
@@ -332,17 +332,18 @@ static int fi_ibv_ep_bind(struct fid *fid, struct fid *bfid, uint64_t flags)
...
 				if (ret) {
 					fastlock_release(&ep->eq->lock);
 					return -errno;
 				}
+			} else {
+				ret = rdma_migrate_id(ep->id, ep->eq->channel);
+				if (ret)  {
+					fastlock_release(&ep->eq->lock);
+					return -errno;
+				}
```
[the remark this answers is not in the document]
Comment (swelch) [reply prov/verbs/src/verbs_ep.c]: Right, I will update the associated commit with this simplification. Thanks for the recommendation.

File: prov/verbs/src/fi_verbs.h (hunk 1 above)
Comment (swelch) [reply prov/verbs/src/fi_verbs.h]: Agreed, I will update the comment in the associated commit. Thanks.

**Comment by swelch** [comment]
@shefty - Pushed updates to address your comments/recommendations. Thanks again.

**Comment by jswaro** [comment]
We've tested this thoroughly with our dev systems.

**Pull request merged**